src/compiler/checker.ts

```diff
@@ -17651,12 +17652,51 @@ namespace ts {
             }
         }
 
+        function getOptionalTypeRecursive(type: Type) {
+            if ((type as ResolvedType).properties) {
+                const rType = type as ResolvedType;
+                const propCount = rType.properties.length;
+
+                if (rType.flags & TypeFlags.Object) {
+                  for (let i = 0; i < propCount; i++) {
+                    const prop = rType.properties[i] as SymbolLinks;
+                    let innerType;
+                    if (prop.bindingElement && prop.bindingElement.initializer) {
+                        innerType = getOptionalTypeRecursive(prop.type);
+                        innerType = getOptionalType(innerType);
+                    }
+                    else if (prop.type.symbol) {
+                        innerType = getTypeOfParameter(prop.type.symbol);
+                    }
+                    else {
+                        innerType = prop.type;
+                    }
+                    prop.type = innerType;
+                    rType.properties[i] = prop as Symbol;
+                  }
+                }
+
+                if (type.flags & TypeFlags.Union || type.flags & TypeFlags.Intersection) {
+                  for (let i = 0; i < propCount; i++) {
+                    const prop = rType.properties[i] as SymbolLinks;
+                    const innerType = getOptionalTypeRecursive(prop.type);
+                    prop.type = innerType;
+                    rType.properties[i] = prop as Symbol;
+                  }
+                }
+                return rType as Type;
+            }
+            else {
+              return type;
+            }
+        }
+
         function getTypeOfParameter(symbol: Symbol) {
             const type = getTypeOfSymbol(symbol);

```

Comment:
I think that you need to follow `getTypeOfSymbol` down to where the type of binding elements is recursively obtained, and change the innermost case, instead of making a post-processing step that introduces `undefined` to the type after the fact.

Reply:
I did exactly that in one of my fix prototypes. It broke a dozen or so of testcases. Furthermore, i don't think this approach is correct. Consider this: getTypeOfSymbol is also used "inside" a function to determine what type a argument has, as well as the signature of the function from the "outside". In this case the two are different. If there is an initializer, the argument in the function should not include undefined, because it is garantueed to never assume this value, but the signature should, because you can safely pass an undefined. We already did that actually, but just for the most "outer level" (see getTypeOfParameter), which is not correct as it has been previously. I agree that this solution (apart from the obvious improvements mentioned) is not perfect, but without changing typechecking on a more fundamental level, i don't see any other way.